File: set6/5/gh_sol.py
import sys
my_input = sys.stdin.readline
special = {
    210: 3,
    220: 4,
    221: 4,
    222: 3,
    225: 3,
    238: 3,
    245: 4,
    247: 5,
    249: 4,
    250: 3,
    256: 3,
    266: 3
}
hh, mm, ss = map(int, my_input().split(':'))
n = int(my_input())
train = []
for i in range(n):
    s, e, ti = my_input().split(' ')
    sec = int(ti.split(":")[0]) * 60 * 60 + int(ti.split(":")[1]) * 60
    train.append(tuple([int(s[1:]), int(e[1:]), sec]))
cur_sec = 3600*hh + 60*mm + ss
arr = [0] * 300
nj = [0] * 300
for t in range(1, 300):
    if t in special:
        arr[t] = special[t] * 60 + 20
    else:
        arr[t] = 2 * 60 + 20
for t in range(1, 300):
    nj[t] = nj[t-1] + arr[t]


def get_duration(station_s, station_e):
    return nj[station_e-1] - nj[station_s-1]


stop = [225, 233, 246, 258, 272]
for i in range(0, 4):
    cur_station = stop[i]
    next_station = stop[i+1]
    mmn_time = 0x3f3f3f3f
    for t in train:
        ts, te = t[0], t[1]
        if te <= cur_station or cur_station < ts:
            continue
        time_start = t[2]
        leave_time = time_start + get_duration(ts, cur_station)
        if cur_sec < leave_time < mmn_time:
            mmn_time = leave_time
    if mmn_time == 0x3f3f3f3f:
        print('-1')
        exit(0)
    cur_sec = mmn_time + get_duration(cur_station, next_station) - 20
if cur_sec >= 86400:
    print('-1')
else:
    print(f'{cur_sec//3600:02d}:{(cur_sec%3600)//60:02d}:{(cur_sec%3600)%60:02d}')

File: set6/5/gh_sol.java
import java.util.*;
import java.lang.*;
import java.io.*;
class Dia{
    private final int []arr = new int[300];
    private final int []njT = new int[300];
    Dia(String []dat, int defaultTime, int stopTime){
        for(int i=1; i<300; i++)
            this.arr[i] = defaultTime + stopTime;

        for(String data : dat){
            String []tokens = data.split("\\s+");
            fill(tokens[0], tokens[2], 20);
            fill(tokens[3], tokens[5], 20);
        }

        for(int i=1; i<300; i++)
            this.njT[i] = this.njT[i-1] + this.arr[i];

    }
    public int getDuration(int s, int e){
        return this.njT[e-1] - this.njT[s-1];
    }
    private void fill(String stationNum, String time, int stopTime){
        int loc = Integer.parseInt(stationNum.substring(1));
        int minute = Integer.parseInt(time);
        this.arr[loc] = 60*minute + stopTime;
    }
}
class Train{
    private final int start;
    private final int end;
    private final int startSecond;
    public Train(String start, String end, String time){
        this.start = Integer.parseInt(start.substring(1));
        this.end = Integer.parseInt(end.substring(1));
        String []token = time.split(":");
        this.startSecond = Integer.parseInt(token[0]) * 60 * 60
                + Integer.parseInt(token[1]) * 60;
    }
    public int getStartStation(){
        return this.start;
    }
    public int getEndStation(){
        return this.end;
    }
    public int getStartSecond(){
        return this.startSecond;
    }
    public String toString(){
        return "[" + this.start + " ~ " + this.end + "]: " + this.startSecond;
    }
}
public class Main {
    public static void main(String []args) throws IOException {
        String []dat = {
                "K210	K211	3	K220	K221	4",
                "K221	K222	4	K222	K223	3",
                "K225	K226	3	K238	K239	3",
                "K245	K246	4	K247	K248	5",
                "K249	K250	4	K250	K251	3",
                "K256	K257	3	K266	K267	3"
        };
        Dia trainDia = new Dia(dat, 120, 20);

        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        String []parse;
        parse = bf.readLine().split(":");
        int hh = Integer.parseInt(parse[0]);
        int mm = Integer.parseInt(parse[1]);
        int ss = Integer.parseInt(parse[2]);
        int curSecond = 60*60*hh + 60*mm + ss;
        int n = Integer.parseInt(bf.readLine());
        Train []tr = new Train[n];

        for(int i=0;i<n;i++){
            parse = bf.readLine().split(" ");
            tr[i] = new Train(parse[0], parse[1], parse[2]);
        }

        int []rideStation = {225, 233, 246, 258, 272};
        for(int i=0; i<4; i++){
            int curStation = rideStation[i];
            int nextStation = rideStation[i+1];
            int mmnTime = 0x3f3f3f3f;
            for(Train t : tr){
                int ts = t.getStartStation();
                int te = t.getEndStation();
                if(te <= curStation || curStation < ts)
                    continue;
                int timeStart = t.getStartSecond();
                int leaveTime = timeStart + trainDia.getDuration(ts, curStation);
                if(curSecond < leaveTime && leaveTime < mmnTime)
                    mmnTime = leaveTime;
            }
            if(mmnTime == 0x3f3f3f3f){
                System.out.println("-1"); return;
            }
            curSecond = mmnTime + trainDia.getDuration(curStation, nextStation) - 20;
            //printTime(curSecond);
        }
        if(curSecond >= 86400)
            System.out.println(-1);
        else
            printTime(curSecond);
    }
    private static void printTime(int sec){
        int h = sec/3600;
        int m = (sec%3600)/60;
        int s = (sec%3600)%60;
        System.out.println(String.format("%02d",h)+":"
                +String.format("%02d",m)+":"+String.format("%02d",s));
    }
}